app/Libraries/Search/BeatmapsetSearch.php

```diff
@@ -129,140 +158,160 @@ public function records()
     {
-        $query->mustNot(['terms' => ['user_id' => $this->params->blockedUserIds()]]);
+        $this->query->mustNot(['terms' => ['user_id' => $this->params->blockedUserIds()]]);
     }
 
-    private function addCreatorFilter(BoolQuery $query, BoolQuery $nested): void
+    private function addCreatorFilter(): void
     {
-        $value = $this->params->creator;
+        $value = $this->includes->creator;
 
-        if (!present($value)) {
-            return;
+        if (present($value)) {
+            $user = User::lookup($value);
+
+            if ($user === null) {
+                $this->addTextFilter('creator', ['creator']);
+            } else {
+                $this->nested->filter(['term' => ['beatmaps.user_id' => $user->getKey()]]);
+            }
         }
 
-        $user = User::lookup($value);
+        $value = $this->excludes->creator;
 
-        if ($user === null) {
-            $this->addTextFilter($query, 'creator', ['creator']);
-        } else {
-            $nested->filter(['term' => ['beatmaps.user_id' => $user->getKey()]]);
+        if (present($value)) {
+            $user = User::lookup($value);
+
+            if ($user === null) {
+                $this->addTextFilter('creator', ['creator'], false);
+            } else {
+                $this->nestedMustNot->should(['term' => ['beatmaps.user_id' => $user->getKey()]]);
+            }
         }
     }
 
-    private function addDifficultyFilter(BoolQuery $nested)
+    private function addDifficultyFilter()
     {
-        if ($this->params->difficulty !== null) {
-            $nested->must(QueryHelper::queryString($this->params->difficulty, ['beatmaps.version'], 'and'));
+        if ($this->includes->difficulty !== null) {
+            $this->nested->filter(QueryHelper::queryString($this->includes->difficulty, ['beatmaps.version'], 'and'));
+        }
+
+        if ($this->excludes->difficulty !== null) {
+            $this->nestedMustNot->should(QueryHelper::queryString($this->excludes->difficulty, ['beatmaps.version'], 'or'));
         }
     }
 
-    private function addExtraFilter($query)
+    private function addExtraFilter()
     {
         foreach ($this->params->extra as $val) {
-            $query->filter(['term' => [$val => true]]);
+            $this->query->filter(['term' => [$val => true]]);
         }
     }
 
-    private function addFeaturedArtistFilter($query)
+    private function addFeaturedArtistFilter()
     {
-        if ($this->params->featuredArtist !== null) {
-            $trackIds = ArtistTrack::where('artist_id', $this->params->featuredArtist)->pluck('id');
-            $query->filter(['terms' => ['track_id' => $trackIds]]);
+        if ($this->includes->featuredArtist !== null) {
+            $trackIds = ArtistTrack::where('artist_id', $this->includes->featuredArtist)->pluck('id');
+            $this->query->filter(['terms' => ['track_id' => $trackIds]]);
+        }
+
+        if ($this->excludes->featuredArtist !== null) {
```

Comment:
I'm not sure where it went wrong but this doesn't seem to work quite right (168 is a-one):

![](https://s.kohaku.love/2025/05/firefox_2025-05-29_21-17-51.png)

![](https://s.kohaku.love/2025/05/firefox_2025-05-29_21-18-47.jpg)

include filter works as expected

![](https://s.kohaku.love/2025/05/firefox_2025-05-29_21-20-16.jpg)